### PR TITLE
Submission rewritten, refreshed, remodulated, restructured.

### PR DESCRIPTION
**Pull request by Neloop**

### CODE REVIEW

File: app/V1Module/presenters/SubmissionsPresenter.php [hunk 1]
```diff
@@ -56,18 +64,30 @@ class SubmissionsPresenter extends BasePresenter {
...
 
-    $submissions = array_filter($this->submissions->findAll(), (function (Submission $submission) {
-      return $this->submissionAcl->canViewDetail($submission);
-    }));
+    $submissions = array_filter($solution->getSubmissions()->getValues(),
+      function (AssignmentSolutionSubmission $submission) use ($solution) {
+        return $this->submissionAcl->canViewEvaluation($solution, $submission);
+    });
+
+    // display only proper data for logged user
```
Comment (Teyras) [inline app/V1Module/presenters/SubmissionsPresenter.php]: I would probably reword this along the lines of "Display only data that the current user can view".

File: app/V1Module/presenters/SubmitPresenter.php [hunk 2]
```diff
@@ -101,6 +107,20 @@ class SubmitPresenter extends BasePresenter {
 
 
   /**
+   * Determine if given user can submit solutions to assignment.
+   * @param Assignment $assignment
+   * @param User|NULL $user
+   * @return bool
+   */
+  public function canReceiveSubmissions(Assignment $assignment, User $user = null) {
```
Comment (Teyras) [inline app/V1Module/presenters/SubmitPresenter.php]: This was moved from the Assignment entity, right? This is certainly a better place, but I would make this method private.

File: app/V1Module/presenters/SubmitPresenter.php [hunk 3]
```diff
@@ -142,7 +162,7 @@ public function actionSubmit(string $id) {
       throw new ForbiddenRequestException();
     }
 
-    if (!$assignment->canReceiveSubmissions($loggedInUser)) {
+    if (!$this->canReceiveSubmissions($assignment, $loggedInUser)) {
       throw new ForbiddenRequestException("User '{$loggedInUser->getId()}' cannot submit solutions for this exercise any more.");
```
Comment (Teyras) [inline app/V1Module/presenters/SubmitPresenter.php]: s/exercise/assignment

File: app/V1Module/router/RouterFactory.php [hunk 4]
```diff
@@ -239,12 +239,13 @@ private static function createReferenceSolutionsRoutes(string $prefix): RouteLis
    */
   private static function createSubmissionRoutes(string $prefix): RouteList {
     $router = new RouteList();
-    $router[] = new GetRoute("$prefix", "Submissions:");
-    $router[] = new GetRoute("$prefix/<id>", "Submissions:evaluation");
+    $router[] = new GetRoute("$prefix/evaluation/<id>", "Submissions:evaluation");
+    $router[] = new GetRoute("$prefix/evaluation/<id>/download-result", "Submissions:downloadResultArchive");
+
     $router[] = new PostRoute("$prefix/<id>", "Submissions:setBonusPoints");
```
Comment (Teyras) [inline app/V1Module/router/RouterFactory.php]: Bonus points should be associated with solutions, not submissions

File: app/V1Module/router/RouterFactory.php [hunk 5]
```diff
@@ -239,12 +239,13 @@ private static function createReferenceSolutionsRoutes(string $prefix): RouteLis
...
+    $router[] = new GetRoute("$prefix/<id>/evaluations", "Submissions:evaluations");
     $router[] = new PostRoute("$prefix/<id>/set-accepted", "Submissions:setAcceptedSubmission");
     $router[] = new DeleteRoute("$prefix/<id>/unset-accepted", "Submissions:unsetAcceptedSubmission");
```
Comment (Teyras) [inline app/V1Module/router/RouterFactory.php]: This also applies to solutions and not submissions.

File: app/V1Module/security/ACL/ISubmissionFailurePermissions.php [hunk 6]
```diff
@@ -2,12 +2,12 @@
...
 use App\Model\Entity\SubmissionFailure;
 
 interface ISubmissionFailurePermissions {
   function canViewAll(): bool;
   function canView(SubmissionFailure $failure): bool;
   function canResolve(SubmissionFailure $failure): bool;
-  function canViewForSubmission(Submission $submission): bool;
-}
\ No newline at end of file
+  function canViewForSubmission(AssignmentSolutionSubmission $submission): bool;
```
Comment (Teyras) [inline app/V1Module/security/ACL/ISubmissionFailurePermissions.php]: I would probably rename this method to canViewForAssignment[Solution]Submission

File: app/V1Module/security/ACL/ISubmissionPermissions.php [hunk 7]
```diff
@@ -2,15 +2,18 @@
 namespace App\Security\ACL;
 
 
-use App\Model\Entity\Submission;
+use App\Model\Entity\AssignmentSolution;
+use App\Model\Entity\AssignmentSolutionSubmission;
 
 interface ISubmissionPermissions {
```
Comment (Teyras) [inline app/V1Module/security/ACL/ISubmissionPermissions.php]: This should be renamed to IAssignmentSolutionPermissions

File: app/V1Module/security/Policies/SubmissionPermissionPolicy.php [hunk 8]
```diff
@@ -2,20 +2,18 @@
 namespace App\Security\Policies;
 
 
-use App\Model\Entity\Submission;
+use App\Model\Entity\AssignmentSolution;
+use App\Model\Entity\AssignmentSolutionSubmission;
 use App\Security\Identity;
 
 class SubmissionPermissionPolicy implements IPermissionPolicy {
```
Comment (Teyras) [inline app/V1Module/security/Policies/SubmissionPermissionPolicy.php]: This should also be renamed.

File: app/model/entity/ReferenceSolutionSubmission.php [hunk 9]
```diff
@@ -80,6 +48,13 @@ public function setEvaluation(SolutionEvaluation $evaluation) {
     $this->evaluation = $evaluation;
   }
 
+  /**
+   * @var Collection
+   * @ORM\OneToMany(targetEntity="SubmissionFailure", mappedBy="referenceSolutionSubmission")
+   */
+  protected $failures;
+
+
```
Comment (Teyras) [inline app/model/entity/ReferenceSolutionSubmission.php]: Remove this empty line, please :D

File: app/V1Module/presenters/SubmitPresenter.php (hunk 2 above)
Comment (Neloop) [reply app/V1Module/presenters/SubmitPresenter.php]: Definitely a good idea :-)

File: app/V1Module/security/ACL/ISubmissionFailurePermissions.php (hunk 6 above)
Comment (Neloop) [reply app/V1Module/security/ACL/ISubmissionFailurePermissions.php]: Will do

File: app/model/entity/ReferenceSolutionSubmission.php (hunk 9 above)
Comment (Neloop) [reply app/model/entity/ReferenceSolutionSubmission.php]: No

File: app/V1Module/presenters/AssignmentSolutionsPresenter.php [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+  /**
+   * @var IAssignmentSolutionPermissions
+   * @inject
+   */
+  public $assignmentSolutionAcl;
+
+
+  /**
+   * Try to get evaluation of given submission. If found, then evaluated it and
```
Comment (krulis-martin) [inline app/V1Module/presenters/AssignmentSolutionsPresenter.php]: typo

File: app/V1Module/presenters/SubmitPresenter.php [hunk 11]
```diff
@@ -101,6 +107,20 @@ class SubmitPresenter extends BasePresenter {
 
 
   /**
+   * Determine if given user can submit solutions to assignment.
+   * @param Assignment $assignment
+   * @param User|NULL $user
+   * @return bool
+   */
+  private function canReceiveSubmissions(Assignment $assignment, User $user = null) {
+    return $assignment->isPublic() &&
```
Comment (krulis-martin) [inline app/V1Module/presenters/SubmitPresenter.php]: I am not completely sure this is right. Take the following scenario. The teacher assigns an exercise. Once the submits begin to arrive, the teacher realizes that there is a mistake, so he/she make the assignment not public, so no more students will get wrong evaluations until the exercise is fixed. Once it is fixed, the patch needs to be tested, preferably on already submitted solutions -- i.e., the teacher needs to re-submit the solutions before making the assignment public again.

Comment (Neloop) [reply app/V1Module/presenters/SubmitPresenter.php]: canReceiveSubmissions() is not used during resubmits.

Comment (krulis-martin) [reply app/V1Module/presenters/SubmitPresenter.php]: OK. My mistake. Let the merging commence!

**Pull request merged**